bin/install

```diff
@@ -1,137 +1,215 @@
 
-    local luaJIT_configure_options="$(construct_luaJIT_configure_options)"
     # set in os_based_configure_options
     # we unset it here because echo-ing changes the return value of the function
     unset ASDF_PKG_MISSING
 
-    if [[ "$OSTYPE" = "darwin"* ]] && [[ -z "$MACOSX_DEPLOYMENT_TARGET" ]]; then
-      export MACOSX_DEPLOYMENT_TARGET=$(/usr/bin/sw_vers --productVersion)
+    if [[ "$OSTYPE" = "darwin"* ]] && [[ -z "${MACOSX_DEPLOYMENT_TARGET-}" ]]; then
+      MACOSX_DEPLOYMENT_TARGET=$(/usr/bin/sw_vers --productVersion)
+      export MACOSX_DEPLOYMENT_TARGET
     fi
 
-    echo "Building luaJIT with options: $luaJIT_configure_options"
-    make amalg $luaJIT_configure_options || exit 1
-    make install $luaJIT_configure_options || exit 1
+    printf "Building luaJIT with options:"
+    printf " %q" "${luaJIT_configure_options[@]}"
+    printf "\n"
+    make amalg "${luaJIT_configure_options[@]}" || exit 1
+    make install "${luaJIT_configure_options[@]}" || exit 1
 
     # luajit beta versions do not symlink luajit to the built binary,
     # so we'll do that manually here, for convenience and also so
     # luarocks can install.
     if [[ $luaJIT_version == *"beta"* ]]; then
-      ln -sf luajit-$luaJIT_version $ASDF_INSTALL_PATH/bin/luajit
+      ln -sf luajit-"$luaJIT_version" "$ASDF_INSTALL_PATH"/bin/luajit
     fi
 
     ##########################################################################
 
-    cd $(dirname $lua_rocks_source_path)
-    tar zxf $lua_rocks_source_path || exit 1
-    cd $(untar_lua_rocks_path $install_type $lua_rocks_version $tmp_download_dir)
+    cd "$(dirname "$lua_rocks_source_path")" || exit 1
+    local source_dir
+    source_dir=$(untar_lua_rocks_path "$install_type" "$lua_rocks_version" "$tmp_download_dir")
+    tar zxf "$lua_rocks_source_path" || exit 1
+    cd "$source_dir"
+
+    local lua_rocks_configure_options=()
+    while IFS='' read -r line; do
+      lua_rocks_configure_options+=("$line")
+    done < <(construct_lua_rocks_configure_options)
 
-    local lua_rocks_configure_options="$(construct_lua_rocks_configure_options)"
     # set in os_based_configure_options
     # we unset it here because echo-ing changes the return value of the function
     unset ASDF_PKG_MISSING
 
-    echo "Building luarocks with options: $lua_rocks_configure_options"
-    ./configure $lua_rocks_configure_options || exit 1
-    make build || make || exit 1
-    make install || exit 1
+    printf "Building luarocks with options:"
+    printf " %q" "${lua_rocks_configure_options[@]}"
+    printf "\n"
+    ./configure "${lua_rocks_configure_options[@]}" || exit 1
+    make bootstrap || exit 1
+
+    # the install leaves some references to the build dir; remove them:
+    local file
+    for file in "$ASDF_INSTALL_PATH"/luarocks/bin/{luarocks,luarocks-admin}; do
+      # this is the path entry that we want to remove
+      local pattern="${source_dir}/src/?.lua;"
```

Comment:
This is really where that path conanicalization helped.

The fragment I wanted to remove looked like this in my case:

```
/private/var/folders/b8/ssbqlklj0h99w3mw1gw6qxpc0000gn/T/luajit-2.1/src/?.lua;
```

(or thereabouts)

But if we looked at the `$source_dir`, it was:

```
/var/folders/b8/ssbqlklj0h99w3mw1gw6qxpc0000gn/T//luajit-2.1/src/?.lua;
```

So the build system (very reasonably) resolved the "real path" and embedded that file temp file path into these files (less reasonable, need to file an issue upstream). To remove the path, I'd need to call `realpath "$source_dir"` _somewhere_.

Instead of doing that only here, it made more sense to run the source directory path through `realpath` at the very beginning. That keeps any mentioning of this path consistent, and gets rid of things like the ugly `//` (due to `$TMPDIR` already ending with a `/` on macOS).

One of those things where there's zero downside, only positives.

Reply:
This is fantastic.